### DISCUSSION TITLE
Questions and chit-chat

### DISCUSSION THREAD

**Discussion by leptun** [vintagepc/MK404#295] [General]
Hi @vintagepc 
Thank you for inviting me to this project. I've read what you wrote so far in readme.md. I have a few questions:
How do you propose that we communicate after you push the first usable source code?
Are there others involved in the project?

**Comment by vintagepc** [comment]
You're welcome!
Right now it's just you and me (I invited wavexx as well). 

I don't have anything in particular lined up for comms, I figured that with a small number of folks we could probably just use issues like a forum thread for notifications and discussion.
Certainly once we get going we can look at alternatives (slack, etc) for casual discussion. 

As this is a private github repo I'm currently limited to 3 invites, so I'm playing it by ear. If you have suggestions for other folks that might be able to contribute I'm happy to hear. The private nature is only temporary though - mostly because it's in the early stages of bring-up and many of the tasks are serial in nature. Once the firmware at least boots, we can "fan out" and it's easier for multiple people to work on tasks like implementing a piece of hardware without either duplicating efforts or stepping on each others' work.

I don't have any expectations on contributions or anything - I invited folks out of interest in the project and not because I had expectations they would contribute :)

In terms of status I'm hoping to spend a bit of time each evening this weekend some of the weekend to see how far I get, and to clean up what I've got into something consistent and maintainable to push it.

**Comment by leptun** [comment]
@vintagepc yes, Wavexx just also saw your invite. I’m thinking that maybe 3d-gussner could also be interested in this. I’ll ask him later.

I’m really curious about the stepper implementation. Imagine emulating MK3 hardware completely. What I mean by this is creating actual “hardware” limits for tmc drivers and loading a custom MBL matrix for testing the code and stuff like that.
Also, the UVLO can be tested in an environment that is much more controlled than cutting power.

**Comment by vintagepc** [comment]
Absolutely!  Once we get this going there's very little we won't be able to do in terms of poking and prodding to set up test cases. 

I'm also happy to invite 3d-gussner if he's interested. I contemplated dracer as well but I don't have a good sense of whether he'd be interested this early since he's got other priorities.

Also, 
PROGRESS! (I traced #11 down to an issue with the serial IO handling. Disabling it in the firmware... it boots!

![image](https://user-images.githubusercontent.com/53943260/78669598-820ec180-78aa-11ea-80bc-3bf8e1cf7158.png)

![image](https://user-images.githubusercontent.com/53943260/78669836-dd40b400-78aa-11ea-818d-72fd7a4c7a00.png)

**Comment by leptun** [comment]
@vintagepc Would it be better if we started smaller? I also want the target to be MK3/S, but maybe it would be easier if we started with MK2.5/S and worked our way up to the other printers.

**Comment by leptun** [comment]
@vintagepc Also, is anything in the emulator persistent?
After a "reboot", do you still have the same EEPROM contents?
And as for the W25X20CL flash, I'd make sure that doesn't erase itself every time we change anything in the code.

And @3d-gussner would also like to join this project.

**Comment by vintagepc** [comment]
Maybe - though I get the impression the main part is soon going to be just getting hardware modules online and I think that after the initial layout learning curve that should go pretty quick for things in our areas of expertiese, where each bit is a relatively straightforward component.  At least in my case my frame of reference is the MK3S and I'm not intimately familiar with the MK2.5S and its bits/pieces. I have more visibility right now since I'm looking at the interfaces and code and mapping it out in my mind, but I think this'll get clearer once I push it up for you guys too. 

I'm definitely not opposed to that though, far from it. 

Any memory is easily made persistent. The program memory is already - for faster testing turnaround I have it reload the firmware/bootloader each time but that can be disabled. The contents are stored as a .bin file on disk (easily hexdumped for inspection) (you could theoretically program it with `objcopy -I hex <firmware.hex> -O binary <flash.bin>`

I've been conservative with my estimates on timing, progress has been progressing faster than I anticipated so I may push something up sooner rather than later. 

Side note, I think this bringup will also reveal some interesting things in the firmware codebase as well, for example https://github.com/prusa3d/Prusa-Firmware/issues/2582 - Not necessarily things that are broken, but just something that someone should take a second look at.

I sent 3d-gussner an invite. I'm actually really excited for where this is going given I tossed it out as a casual idea; the mind still boggles that I even managed to get it to boot to the main marlin menu in the span of a few days. 👀

**Comment by vintagepc** [comment]
Also I think it would be ultra-cool if we could get this wired up enough to the point we could even graphically simulate a print. Would Prusa be the first to have a fully virtual 3d printer? Might even have implications for something like P'licer giving you a toolpath or print animation!

Lofty goals, I know. So many ideas, so little time!

**Comment by leptun** [comment]
I was also thinking about that, but I think it is a bit too early to even start thinking about those implementations. I agree with you that a live 3D model of the "printer" would be quite useful for some functions.
Btw, what's the performance of the emulation? Is it horribly slow?
In my case, the fans ramp up to max as soon as I start the hd44780 example in the simAvr source...

**Comment by vintagepc** [comment]
Yeah, that's definitely a looooooooooooong way off.
On my laptop (from 2013) it feels a bit more laggy than the real thing, but I often have program counter logging on since I'm still debugging early boot stages.

I'm pretty sure the examples are not well optimized, I think the display one basically has a while(true) with no sleep cycles in it somewhere... Also, that example enables VCD traces by default so it's logging a ton of data every 10 usec, IIRC)

**Comment by leptun** [comment]
You say it feels a bit more laggy... Would it even be possible in the future to speed up the runtime for generating some of the weirder bugs that don't happen instantly? (Aka fast-forward)

**Comment by vintagepc** [comment]
I think so. I'm almost certain some of the lag comes from unoptimized things like display draws, I am definitely not a GL expert, I think it updates anytime there's a display write, for example. No reason that couldn't be changed to delay slightly in order to aggregate several character writes.
 Once we get a proof of concept I think we should be able to get improvements. I haven't gotten to the point of trying to profile stuff yet.

**Comment by leptun** [comment]
@vintagepc Just saying, when the LCD buffer will get merged into the MK3 branch, we could make it even more lightweight. Just read the buffer in memory for the characters. Only the custom characters should be a bit more interesting.
Btw, in my latest push to that PR I also created a "banking" implementation for the custom characters (since we have 11 chars that must fit in 8 cells). With the new implementation, each and every custom character has a dedicated uint8_t value. This means that we could even just edit the font of the drawing program to have our custom characters in the 0x00-0x0F range.

**Comment by vintagepc** [comment]
Oh, cool! I contemplating implementing the custom character upload to-spec on the reference display. But as noted in #4 there are some issues with the stock implementation to address first with regards to framing/memory locations.

**Comment by 3d-gussner** [comment]
Thanks for inviting me.
I am Waldi aka 3d-gussner.

**Comment by vintagepc** [comment]
Welcome aboard!
BTW, if folks have any particular issues/hardware they are interested in working on then please feel free to comment on or claim them. I'm hoping to get my exploratory hackery cleaned up in the next few days and then we'll be off to the races!

**Comment by leptun** [comment]
@vintagepc I’m a bit new to coding outside embedded platforms... How exactly do you make it “run”. I know that is a basic thing, but I’d like to have a basic understanding of what is going on. After that, I can begin to understand how your code works and move on to the W25X20CL task.

**Comment by vintagepc** [comment]
Ah, sure. Let me finish pushing up the rest of my assortment of stuff. I just finished templating a basic SPI dummy class to see if I could get it to print out the bytes it's receiving from the AVR. 
I'll follow up with a readme.

**Comment by vintagepc** [comment]
I've bumped up what I have so far along with a few readmes and whatnot. 
Let me know if you have any questions/issues getting started!

**Comment by vintagepc** [comment]
Observation: Partway through the TMC2130 work I think that's part of what's causing it to be sluggish. It's constantly detecting crashes and trying to rehome (even if I cancel the setup wizard)

**Comment by vintagepc** [comment]
Github now has unlimited collaborators on private repos - so if you can think of other folks that should be invited to preview/contribute to the project, please don't hesitate to let me know!

**Comment by wavexx** [comment]
What about @dracer :)

**Comment by leptun** [comment]
@DRracer not @dracer

**Comment by wavexx** [comment]
Yeah sorry :)

**Comment by vintagepc** [comment]
Invite sent!

**Comment by leptun** [comment]
@wavexx I've done this mistake in the past. The only way I noticed the mistake was that the text wasn't capitalized automatically.
RIP @dracer for the spam he probably gets whenever someone mistypes @DRracer's name

**Comment by vintagepc** [comment]
"Who the hell are all these random people mentioning me on 3d printing stuff? " :rofl:

**Comment by DRracer** [comment]
oh, now I understand ;) . Thanks for inviting me into the simulator.

**Comment by vintagepc** [comment]
Welcome aboard!

**Comment by DRracer** [comment]
@vintagepc @leptun please update the doc a bit, it took me a while figuring out I had to clone the simavr from source, copy the cloned MK3SIM into examples/board_einsy (!).
And - which branch is the most recent one? I'm having a compile issues, which has probably been already solved elsewhere:
`Einsy.c:381:50: error: ‘TIMER_IRQ_OUT_PWM2’ undeclared (first use in this function); did you mean ‘TIMER_IRQ_OUT_PWM1’?`

**Comment by leptun** [comment]
@DRracer The "documentation" for setting up the build env is this: https://github.com/vintagepc/MK3SIM/issues/26#issuecomment-613130844
We'll write a real documentation later.

**Comment by vintagepc** [comment]
I'll update the 'getting started` file with that comment's contents for starters.

**Comment by vintagepc** [comment]
Done. Sorry about that- we've been moving pretty fast over the last two weeks so the docs haven't really kept up.

**Comment by DRracer** [comment]
@vintagepc no problem, I know this is an early sprint ;)

**Comment by vintagepc** [comment]
This is just too cool not to share...
https://vimeo.com/409042050

Edit - also:

![vokoscreen-2020-04-17_18-58-17](https://user-images.githubusercontent.com/53943260/79620422-d6e9dd80-80dd-11ea-944b-fbb778201bbf.gif)
(Gif is slower than reality, just didn't transcode well)

**Comment by vintagepc** [comment]
Today is a big day... There are definitely still some issues to resolve and features to implement... but I am pleased to announce I've just successfully completed the calibration wizard on the simulator! (Edit: whoa, has it only been a little over two weeks??!?)

![image](https://user-images.githubusercontent.com/53943260/80157549-6ab32200-8594-11ea-8d06-d17b0963f96d.png)

![image](https://user-images.githubusercontent.com/53943260/80157652-ab12a000-8594-11ea-8c07-5ad99be7d755.png)

![image](https://user-images.githubusercontent.com/53943260/80157780-ef9e3b80-8594-11ea-9a96-d0d2c7c98e7e.png)

![image](https://user-images.githubusercontent.com/53943260/80157896-3429d700-8595-11ea-923e-9ff9a21caa7e.png)

![image](https://user-images.githubusercontent.com/53943260/80157964-63404880-8595-11ea-9bfe-55668a0d4807.png)

**Comment by DRracer** [comment]
@vintagepc great screenshots - I'm slowly catching up with your progress.
May be a silly question: is it normal that it takes literally minutes before the "printer" gets from the initial screen?

**Comment by vintagepc** [comment]
> @vintagepc great screenshots - I'm slowly catching up with your progress.
> May be a silly question: is it normal that it takes literally minutes before the "printer" gets from the initial screen?

It starts pretty quick for me, maybe about five seconds between starting and the idle info screen being up. My computer's from 2013, so it's not horribly demanding in terms of hardware requirements, I think.

Do you have some details on how you're running it? (hardware, VM, etc) - perhaps that can provide some insight as to what's up. I know some earlier builds were definitely sluggish.

**Comment by DRracer** [comment]
@vintagepc physical machine intel Core2 quad @2GHz running Kubuntu 19.10 ... I made a clean build a few minutes ago. It is just the start, then the simulated printer behaves like a real thing in terms of speed

**Comment by vintagepc** [comment]
Hmm, are you running the bootloader with the -b argument? There are some known glitches with the delay timer that might be part of the issue (#75). Based on the docs from Atmel there isn't a substantial difference between FC and PC pwm modes but it seems to be enough to alter the time tracking. (PC PWM isn't implemented in SimAVR)

**Comment by vintagepc** [comment]
Additional note: I created a few projects we can start using to track the state of various goals/milestones. 

Please feel free to categorize issues as you see fit, the more organized things are the better the end result will be :1st_place_medal:

**Comment by DRracer** [comment]
With or without -b it really doesn't make a difference, it takes about a minute to get into main screen. May be an unoptimized build of the Einsy.elf?

Projects: OK, I'll look into it

**Comment by vintagepc** [comment]
Interesting. I wonder if this is a variant of #59 ... I've noticed the opposite a few times where first boot is fast but subsequent boots seem to run more slowly. The MMU was terribly slow for me until I implemented the buttons because it thought someone was holding down the left button all the time - so I'd not be surprised if were missing a part of something that, when uninitialized can result in odd behaviour. 

Edit: Also, it might help to do a fresh pull of the parent simavr `worktree_test` branch I have and rebuild that with `make build-simavr` to make sure you're up to date with tweaks I've pushed there.

**Comment by leptun** [comment]
@vintagepc Could you please add another person to the project?
Try Miroslav.majer@prusa3d.cz . There's a chance that it might not work.

**Comment by vintagepc** [comment]
sure, I'll give it a go

**Comment by vintagepc** [comment]
Looks like it did not find a github account and sent an invite to his email address :+1:

**Comment by DRracer** [comment]
@vintagepc please contact me via email, we'd like to send you some present for your work on FW 3.9.0.

**Comment by vintagepc** [comment]
How do folks feel about an expanded private beta now that all of the tasks in "Functional MK3S" are complete? Anyone specific come to mind?

**Comment by 3d-gussner** [comment]
I think this would be a great tool for Scott from the Marlin team.
And Gina from Octoprint, not sure if she has the time but could imagine that she would know some developers who would be very interested in this.

**Comment by wavexx** [comment]
On Tue, Jun 23 2020, vintagepc wrote:
> How do folks feel about an expanded private beta now that all of the
> tasks in "Functional MK3S" are complete? Anyone specific come to mind?

Hard to point out anyone specific, but I second @thinkyhead (Scott) for
sure.

I wonder why the "private" beta started in the first place? ;)

**Comment by leptun** [comment]
Isn’t it better when you surprise people with a finished project? That’s what I would personally do. (Hint for a project I’m working on)

**Comment by vintagepc** [comment]
Yes on the last point... I've seen that more general users often lose interest if you share a project too early because they don't have time to deal with the constant updates or things being not functional yet. 

I'll reach out to Scott and Gina :-)

**Comment by wavexx** [comment]
It's a good point. I generally simply don't announce it ;)

**Comment by vintagepc** [comment]
Welcome aboard to @foosel 😄 🐙

**Comment by foosel** [comment]
Hi there! Will have to give this a test soon, looks awesome!

**Comment by 3d-gussner** [comment]
@foosel moin moin, I used it very often to troubleshoot things and used Octoprint for communication. We could troubleshoot few issues in the Prusa Firmware with Octoprint.
Nice thing is that @vintagepc made it better and better.
New boards that are used on different printers are now possible and this opens a new world for testing and development.

**Comment by wavexx** [comment]
Btw, I know on linux simavr creates a tty. I have zero experience on windows, but does it create a virtual com port?
AKA, can you attach octoprint to it? ;)

**Comment by vintagepc** [comment]
You have to build it with Cygwin on windows. I imagine that if you were to make a virtual COM port in Win you could pipe it through to the PTY to get a connection.  I think the project (and simAVR, more importantly) use too much underlying linux stuff to ever be able to build a native windows version.

**Comment by vintagepc** [comment]
Welcome aboard to @thinkyhead!

**Comment by vintagepc** [comment]
I know I've been saying this for a while but I think we've got all the major functionality out of the way... I think we are in good shape to flip the switch to public in the very near future once we wrap up the MK3 variant work! :tada:

**Comment by vintagepc** [comment]
We are public! 

Note this decoupled the previous fork(s) you may have had to private repos, so you will need to create a new fork and/or rebase your changes onto the current repo.

**Comment by leptun** [comment]
Wonderful news! Where are you planning to make a post about the project?

**Comment by vintagepc** [comment]
I haven't thought that far ahead, I figure I might just let it gain its own momentum. I might post about it in the PR forums. 

(I don't do social media like reddit etc, so my exposure avenues are somewhat limited). 

Who knows, maybe someday we'll get featured on a Prusa livestream or blog post... :rofl: 

@DRracer _(wink wink nudge nudge)_

**Comment by DRracer** [comment]
@vintagepc great news! I'll talk to our Content team about the (now public) MK3SIM, in fact I already got some questions about its state and whether it will get public one day.
The best timing to promote your huge effort is with FW 3.9.1 and largely repaired Linear Advance 1.5, which wouldn't be possible without the SIM and direct insight into what is going on in the algorithms during printing. @wavexx , what do you think?

**Comment by DRracer** [comment]
@vintagepc - a short notice:
- cloned with submodules
- apt install libelf-dev libsdl-dev freeglut3-dev
- opened CMakeLists.txt in QtCreator 4.13b1
- compiled like a charm with gcc-10.1.0 ;)
It runs, however I'm still having problems with the boot process taking about a minute. The "P" and "E" letters are rotating, but nothing is happening. It is standing at `pBoard->WaitForFinish();` and some other threads are possibly waiting for UART. At least I can debug it nicely with full comfort in QtCreator ;) . I'll let you know if I discover something (if @leptun isn't faster as always :) ).

**Comment by vintagepc** [comment]
> @vintagepc - a short notice:
> 
> * cloned with submodules
> * apt install libelf-dev libsdl-dev freeglut3-dev
> * opened CMakeLists.txt in QtCreator 4.13b1
> * compiled like a charm with gcc-10.1.0 ;)
>   It runs, however I'm still having problems with the boot process taking about a minute. The "P" and "E" letters are rotating, but nothing is happening. It is standing at `pBoard->WaitForFinish();` and some other threads are possibly waiting for UART. At least I can debug it nicely with full comfort in QtCreator ;) . I'll let you know if I discover something (if @leptun isn't faster as always :) ).

Interesting, please feel free to start a separate issue thread so we can debug this without other messages getting interleaved. 

(You did load a firmware to flash at least once with a -f flag or PS/avrdude, right? Otherwise you might just be bootlooping the bootloader. Don't think this is the case though, sounds like it does eventually finish the boot process)

The wait state sounds about as expected, WaitForFinish() is basically a thread_join that returns when the AVR shuts down. Most of the other threads are indeed just serial helpers. 

It might be useful in this case to inspect the avr->pc value while it's bootlooping and reference that back to an objdump of the .afx file to see in what function it's stuck. (Board.h::RunAVR()). See `FirmwareDebug.md` for info how to make the dump)

**Comment by DRracer** [comment]
@vintagepc sorry, my bad, no firmware (parameter -f). Now it runs well, thanks for your hint.

**Comment by vintagepc** [comment]
No worries. I added a note to the README to that extent, you weren't the first, and won't be the last ;)

**Comment by wavexx** [comment]
On Sun, Jul 26 2020, DRracer wrote:
> @wavexx , what do you think?

I think it would be awesome if possible!

**Comment by 3d-gussner** [comment]
https://twitter.com/MarlinFirmware/status/1287259911303442432

**Comment by vintagepc** [comment]
Heh, that explains the giant spike in traffic yesterday 😁

**Comment by vintagepc** [comment]
I've enabled a wiki for the project, seems like a useful addition as a place to document technical info as required. 

Feel free to add/edit as you see fit, hopefully it will morph into something with a clearer direction as we find uses for it. :)

**Comment by vintagepc** [comment]
Welp, I posted about it in the forums. 
Time to brace for the masses :rofl:

**Comment by GilesBathgate** [comment]
Hi, I tried out this project. I couldn't get selftest working in various scenarios, or calibrateXYZ, I can log some issues for these. My question is have you considered being able to simulate hardware failures or construction issues such as slipping belts, bed/frame skew, or an unlevel heatbed?

**Comment by vintagepc** [comment]
There are quite a few fault conditions that can be simulated, yes - such as motor stalls, thermal runaway/thermistor shorts, custom bed mesh, etc - Have a look at the scripting options for a given printer:
https://github.com/vintagepc/MK404/wiki/Scripting

A common problem with selftest and XYZ calibration is forgetting to toggle the presence of the steel sheet (key `y` ) so that the printer can find the calibration points. 

Does that help?

**Comment by GilesBathgate** [comment]
@vintagepc it seems to be clang. When compiled with gcc/g++ the selftest works. I thought this might be due to #339 since clang could be fussier about the attribute, and also this post seems to suggest it only works when declared after the struct keyword https://stackoverflow.com/questions/13688625/how-to-declare-packed-struct-without-padding-for-llvm/25532933 however changing these doesn't seem to help.

**Reply by vintagepc** [reply]
Oh that's annoying - because from past experience I know if the attribute is anywhere other than where it is currently, it will fail to pack the struct correctly in Cygwin/GCC on Windows - hence why I placed it after. 

Easy enough to check by evaluating some sizeof() calls to see if the structure is the correct size without byte-aligned padding. 

At what stage is your selftest failing?

**Reply by GilesBathgate** [reply]
> Oh that's annoying - because from past experience I know if the attribute is anywhere other than where it is currently, it will fail to pack the struct correctly in Cygwin/GCC on Windows - hence why I placed it after.

I did some tests with sizeof on godbolt,  neither clang 12, nor GCC seem to care where it's placed or whether it has `__padding__` vs `padding`, so if it was a compiler bug it's fixed.

> Easy enough to check by evaluating some sizeof() calls to see if the structure is the correct size without byte-aligned padding.

I guess some unit tests could rule it out, but I think its something else.

> At what stage is your selftest failing?

It gets through to the fan spinning yes/no step, sometimes it will get to the second yes/no, but as soon as the stepper motor tests starts it hangs.

**Reply by vintagepc** [reply]
IIRC this was specifically on Cygwin builds - Linux builds with GCC worked just fine. 

Re: selftest - the entire sim hangs? interesting. What happens if you configure cmake with -DENABLE_THREAD_SANITY=1 (or the clang equivalent if it doesn't use -fsanitize=threads) and build/run to the point of failure? That might highlight the issue if it's related to bad memory accesses

**Reply by GilesBathgate** [reply]
I added -DENABLE_THREAD_SANITY=1 which slowed things down a lot. When I say it hangs, the LCD just freezes, but the P logo spins.
![image](https://user-images.githubusercontent.com/444143/143589445-850d7f36-a818-4c9a-bc1f-89e0e4163934.png)

**Reply by vintagepc** [reply]
BTW you are more than welcome to create a new discussion topic for this, the only reason this thread exists is I migrated it from an issue that was made before discussions were a thing. :)

**Reply by vintagepc** [reply]
hmm, interesting. So the sim isn't stalled or crashing, it seems like it's just not responding to the input . I see you tried the arrow keys too, those did nothing either?

**Reply by GilesBathgate** [reply]
I prefer the keys, but I tried the mouse wheel too.

**Reply by GilesBathgate** [reply]
Sorry, you wanted to know if the cursor on the LCD moves when you press the keys? No, it freezes, the `>` is stuck.

**Reply by vintagepc** [reply]
Thanks, 

Hmm...  the blank lines in the terminal after the output are interesting. That almost seems like the focus is wrong and the enter keypresses are going to the terminal instead of the application window.

**Reply by vintagepc** [reply]
What version of clang are you using? I'll see if it's available for my distro and see if I can reproduce the issue locally

**Reply by GilesBathgate** [reply]
```
$ clang -v
Ubuntu clang version 12.0.0-3ubuntu1~21.04.2
Target: x86_64-pc-linux-gnu
Thread model: posix
```

**Reply by GilesBathgate** [reply]
> The blank lines in the terminal after the output are interesting. That almost seems like the focus is wrong and the enter keypresses are going to the terminal instead of the application window.

Thats my bad, I was just trying to get anything to respond.
